File: SILAM/test_proj/test_proj.f90
program test_proj

use,intrinsic:: ISO_C_BINDING, only: c_int, c_char, c_ptr, c_null_ptr, &
        & c_associated, c_size_t, c_loc, c_funptr, c_null_funptr,c_sizeof
implicit none

interface
  function proj_context_create() bind(C,name='proj_context_create')
    use iso_c_binding
    implicit none
    type(c_ptr) :: proj_context_create
  end function
  subroutine proj_context_destroy(ctx) bind(C,name='proj_context_destroy')
    use iso_c_binding
    implicit none
    type(c_ptr),value :: ctx
  end subroutine
  function proj_create_crs_to_crs(ctx,s_srs,t_srs,area) bind(C,name='proj_create_crs_to_crs') 
    use iso_c_binding                                                                         
    implicit none                                                                             
    type(C_PTR)            :: proj_create_crs_to_crs
    character(kind=C_CHAR) :: s_srs(*)
    character(kind=C_CHAR) :: t_srs(*)
    type(C_PTR), value     :: ctx, area
  end function
  subroutine proj_destroy(PJ) bind(C,name='proj_destroy')
    use iso_c_binding
    implicit none
    type(c_ptr), value :: PJ
  end subroutine
  function proj_trans_generic(pj,dir,x,sx,nx,y,sy,ny,z,sz,nz,t,st,nt) bind(C,name='proj_trans_generic')
    use iso_c_binding
    implicit none
    integer(c_int)       :: proj_trans_generic
    type(c_ptr), value   :: pj
    integer(c_int),value :: dir 
    type(c_ptr),value    :: x,y,z,t
    integer(c_long),value:: sx,sy,sz,st
    integer(c_int),value :: nx,ny,nz,nt
  end function
end interface

!MAIN:
integer :: i
character (len=*), parameter ::     lonlat_proj4="+proj=longlat +ellps=WGS84 +datum=WGS84 +no_defs"
character (len=*), parameter :: hirlam_rll_proj4="+proj=ob_tran +o_proj=longlat +o_lon_p=0 +o_lat_p=30 +lon_0=0"
character (len=*), parameter ::      lcc_example="+proj=lcc +lon_0=-90 +lat_1=33 +lat_2=45"

integer, parameter :: np=4
real(8)     :: lats(np),lons(np)

type(c_ptr) :: CTX = c_null_ptr
type(c_ptr) ::  PJ = c_null_ptr

!corners of EU meteo in projected lon-lat
lons(1:np) = (/-26., -26.0, 40.0,  40.0/)
lats(1:np) = (/-35.,  22.5, 22.5, -35.0/)

CTX=proj_context_create() 
PJ=proj_create_crs_to_crs(CTX, lonlat_proj4, hirlam_rll_proj4, C_NULL_PTR)
    
   call latlon2proj(PJ, lons, lats, np, "PJ_IDENT")
   
   print*,("rotated-pole coordinates lon, lat")
   print*,(hirlam_rll_proj4)
   do i=1,np
     print*,i, lons(i),lats(i)
   enddo
   
   call latlon2proj(PJ, lons, lats, np, "PJ_INV")
   
   print*,("Same in WGS84 coordinates lon, lat")
   print*,(lonlat_proj4)
   do i=1,np
     print*,i, lons(i),lats(i)
   enddo
   
   call latlon2proj(PJ, lons, lats, np, "PJ_FWD")
   
   print*,("And back! lon, lat")
   print*,(hirlam_rll_proj4)
   do i=1,np
     print*,i, lons(i),lats(i)
   enddo

   print*,("Done")

!Clean:
call proj_destroy(PJ)
call proj_context_destroy(CTX)

contains

subroutine latlon2proj(PJ,x,y,np,direction) 
  implicit none
  integer             :: i, status
  type(c_ptr)         :: PJ
  integer, intent(in) :: np
  real(8), dimension (:), target, intent(inout) :: x, y
  character(*)        :: direction !PJ_FWD,PJ_INV,PJ_IDENT
  integer :: dir
  integer(8) :: sp

  if      ( trim(direction) == "PJ_FWD"  ) then
     dir=1;
  else if ( trim(direction) == "PJ_IDENT") then                     
     dir=0;
  else if ( trim(direction) == "PJ_INV"  ) then                   
     dir=-1;
  endif

  sp=c_sizeof(x(1))
  status=proj_trans_generic( PJ, dir,              &
                           & c_loc(x(1)),sp, np,   &
                           & c_loc(y(1)),sp, np,   &
                           & c_null_ptr ,sp, 0,    &
                           & c_null_ptr ,sp, 0)  
end subroutine



end program

File: SILAM/test_proj/test_proj.c
#include <stdio.h>
#include <string.h>
#include <proj.h>
//#include <proj_api.h>

#define d_pi 3.14159265359

int main() {
    int i,status;    
    double drad_to_deg = 180.0/d_pi;
    double ddeg_to_rad = d_pi/180.0;

    char *lonlat_proj4="+proj=longlat +ellps=WGS84 +datum=WGS84 +no_defs";
    char *hirlam_rll_proj4 ="+proj=ob_tran +o_proj=longlat +o_lon_p=0 +o_lat_p=30 +lon_0=0";
    char *lcc_example="+proj=lcc +lon_0=-90 +lat_1=33 +lat_2=45";

    PJ *P;			//define transformation/projection object
    //PJ_COORD c[4], c_out[4];	//define coordinates cin and cout

    const int np=4;
    double lons[4] = {-26.0,-26.0, 40.0,  40.0};
    double lats[4] = {-35.0, 22.5, 22.5, -35.0};
    //CTX = proj_context_create();	//init context.
    //ini proj/transformation
    //P = proj_create_crs_to_crs(PJ_DEFAULT_CTX,lonlat_proj4,hirlam_rll_proj4,NULL);
    //P = proj_create_crs_to_crs(0,lonlat_proj4,hirlam_rll_proj4,0);
    //P = proj_create_crs_to_crs(0,lonlat_proj4,hirlam_rll_proj4,0);
    P = proj_create_crs_to_crs(NULL,lonlat_proj4,hirlam_rll_proj4,NULL);
    
    ////ini coordinates
    //c[0].lpzt.lam=-26.0; //*ddeg_to_rad;
    //c[0].lpzt.phi=-35.0; //*ddeg_to_rad;
    //c[1].lpzt.lam=-26.0; //*ddeg_to_rad;
    //c[1].lpzt.phi= 22.5; //*ddeg_to_rad;
    //c[2].lpzt.lam= 40.0; //*ddeg_to_rad;
    //c[2].lpzt.phi= 22.5; //*ddeg_to_rad;
    //c[3].lpzt.lam= 40.0; //*ddeg_to_rad;
    //c[3].lpzt.phi=-35.0; //*ddeg_to_rad;
    //
    ////(1) (ll_WGS84) -> Hirlam_rll .
    //printf("%s\n","rotated-pole coordinates lon, lat");
    //printf("%s\n",hirlam_rll_proj4);
    //for (i=0;i<=3;i++){
    //    c_out[i] = proj_trans(P, PJ_IDENT, c[i]);
    //    printf("%-3d %20.15f %20.15f\n", i+1, c_out[i].lpzt.lam, c_out[i].lpzt.phi);
    //}
    ////(2) Hirlam_rll -> (ll_WGS84)
    //printf("%s\n","Same in WGS84 coordinates lon, lat");
    //printf("%s\n",lonlat_proj4);
    //for (i=0;i<=3;i++){
    //
    //	c[i] = proj_trans(P, PJ_INV, c_out[i]);
    //    printf("%-3d %20.15f %20.15f\n", i+1, c[i].lpzt.lam, c[i].lpzt.phi);
    //}
    ////(3)((ll_WGS84) -> Hirlam_rll
    //printf("%s\n","And back! lon, lat");
    //printf("%s\n",hirlam_rll_proj4);
    //for (i=0;i<=3;i++){
    // 	
    // 	c_out[i] = proj_trans(P, PJ_FWD, c[i]);
    //    printf("%-3d %20.15f %20.15f\n", i+1, c_out[i].lpzt.lam, c_out[i].lpzt.phi);                                                                      
    //}
    //printf("%s\n","Done");

    
    printf("================\n%s\n================\n"," WORKARROUND");
    printf("size of double: %ld",sizeof(double));
    //(1) (ll_WGS84) -> Hirlam_rll .
    printf("%s\n","rotated-pole coordinates lon, lat");
    printf("%s\n",hirlam_rll_proj4);
    status=proj_trans_generic ( P, PJ_IDENT, lons, sizeof(double), np, lats, sizeof(double), np, 0, 0, 0, 0, 0, 0 );
    for (i=0;i<np;i++){
        printf("%-3d %20.15f %20.15f\n", i+1, lons[i],lats[i]);                                                                      
    }
    //(2) Hirlam_rll -> (ll_WGS84)
    printf("%s\n","Same in WGS84 coordinates lon, lat");
    printf("%s\n",lonlat_proj4);

    status=proj_trans_generic ( P, -1    , lons, sizeof(double), np, lats, sizeof(double), np, 0, 0, 0, 0, 0, 0 );
    //status=proj_trans_generic ( P, PJ_INV, lons, sizeof(double), np, lats, sizeof(double), np, 0, 0, 0, 0, 0, 0 );
    for (i=0;i<np;i++){
        printf("%-3d %20.15f %20.15f\n", i+1, lons[i],lats[i]);                                                                      
    }
    //(3)((ll_WGS84) -> Hirlam_rll
    printf("%s\n","And back! lon, lat");
    printf("%s\n",hirlam_rll_proj4);

    status=proj_trans_generic ( P,  1    , lons, sizeof(double), np, lats, sizeof(double), np, 0, 0, 0, 0, 0, 0 );
    //status=proj_trans_generic ( P, PJ_FWD, lons, sizeof(double), np, lats, sizeof(double), np, 0, 0, 0, 0, 0, 0 );
    for (i=0;i<np;i++){
        printf("%-3d %20.15f %20.15f\n", i+1, lons[i],lats[i]);                                                                      
    }
    printf("%s\n","Done");

   // Clean up 
   proj_destroy(P);

    return 0;
}
